README.md
### P-System-Core
 

### 开发您自己的项目，也可以使用以下开源框架O(∩_∩)O 

[Yii](https://www.yiiframework.com/) | 
[ThinkPHP](https://www.thinkphp.cn/) | 
[CodeignIter](https://codeigniter.com/) | 
[Laravel](https://laravel.com/) | 
[FuelPHP](https://www.fuelphp.com/) | 
[Nette](https://nette.org/) | 
[CakePHP](https://cakephp.org/)


[phalcon](https://phalcon.io/zh-cn) | [yar](https://github.com/laruence/yar)



###  核心部分，无需修改代码。

使用时请使用 `php-start` 开始项目，不能直接使用此扩展包！ 

### [模板](https://github.com/kalimatas/php-liquid)

~~~
$data['products'] = [
  [
    'name'=>'产品名称',
    'price'=>'99.99',
    'description'=>'描述',
  ]
];
echo liquid('demo',$data);
~~~

### 依赖

https://github.com/predis/predis

https://github.com/firebase/php-jwt

https://phpspreadsheet.readthedocs.io/en/latest/

https://github.com/vlucas/valitron

https://medoo.in/

https://uuid.ramsey.dev/en/stable/rfc4122/version1.html


[guzzlephp7]https://docs.guzzlephp.org/en/stable/quickstart.html

[phpseclib2]http://phpseclib.sourceforge.net/


### 如使用 mongodb 请使用

~~~
composer require mongodb/mongodb
~~~

https://www.mongodb.com/docs/php-library/current/tutorial/crud/

 


### 插件forum中使用视图文件

router.php代码演示，对应的目录有两个：
1. /theme/default/forum/tpl/index.php
2. /plugins/forum/tpl/index.php

~~~
global $router; 
$router->get("/forum",function(){
    view("forum@tpl/index"); 
});

~~~ 

#### JWT 
*****
**一般会员**
接口调用以下函数。默认如果用户未登录会返回JSON错误。
~~~
api($show_error = true)
~~~
有时我们需要不抛出错误，使用以下形式
~~~
$api = api(false);
$user_id = $api['user_id'];
~~~

*****
**管理员**
**接口是否是管理员**
~~~
$api = api_admin();
~~~
要求请求中含有`is_admin`参数，且有值。



> 至于跨域问题，接口默认全部跨域。参考代码。无需自行添加。
~~~
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Credentials:true'); 
header("Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization");
header('Access-Control-Allow-Methods: GET, POST, PUT,DELETE,OPTIONS,PATCH');
header('X-Powered-By: WAF/2.0');
if(isset($_SERVER['REQUEST_METHOD']) && $_SERVER['REQUEST_METHOD'] == 'OPTIONS'){ 
    exit;
}
~~~




#### 一些内置函数 
命令行中，防止重复执行
~~~
if(!is_cli()){
	exit('RUN IN CLI');
}
set_time_limit(-1); 
ini_set('memory_limit','1024M');
cli_prevent_duplication($argv, $cmd = 'php io.php');
~~~
判断平台管理员是否登录，未登录自动跳到登录页
~~~
check_admin_login();
~~~ 
写日志
~~~
write_log($msg,$is_error = false)
~~~
生成条形码
~~~
$bar = new lib\Barcode;
echo $bar->display(date("YmdHis"));exit;
~~~
发送邮件,配置在 `config.ini.php`中。
~~~
\lib\Mail::send([
    'from'   => 'ddd@163.com',
    'to'     => $data['email'],
    'subject'=> "标题, 
    'html'   => "<p>内容</p>"
   ]);
~~~
取位置lat,lng
~~~
https://lbs.qq.com/dev/console/key/setting
lib\Map::tx('地址')  腾讯地图
lib\Map::gaode($address, $show_full = false)高德地图
~~~

最近几天
~~~
lib\Time::neerDays($num = 5,$separate = "")
~~~
返回最近几月
~~~
lib\Time::neerMonths($num = 5,$separate = "")
~~~
返回最近几年
~~~
lib\Time::neerYears($num = 5,$separate = "")
~~~
取今日、本周、本月、本年、昨日、上周、上月、上年
~~~
lib\Time::get($key, $date_format = false)
~~~
返回多少岁
~~~
lib\Time::age($bornUnix)
~~~
计算时间剩余　
~~~
lib\Time::less_time($a, $b = null)
~~~
 
返回图片信息
~~~
$file = PATH.'../demos/doc/1.jpg';  
$pdf = new \lib\Img($file); 
//取长宽

print_r($pdf->getWh());
//2是横版，1是竖版
print_r($pdf->getDimensionsType()); 
~~~
多少时间之前
~~~
timeago($time)
~~~

设置配置
~~~
set_config($title,$body)
~~~
优先取数据库，未找到后取配置文件
~~~
get_config($title)
~~~
XML 转成 数组
~~~
xml_to_array($xml)
~~~
数组转xml 
~~~
array_to_xml($arr)
~~~
AES加密
~~~
$config['aes_key'] = "123456";
$config['aes_iv']  = md5('app_sun');


$token = urlencode(aes_encode($d)); 
~~~
返回两个时间点间的日期数组
~~~
get_dates($start, $end)
~~~
当前时间是周几
~~~
get_current_week($date)
~~~
多语言翻译

~~~
lang('welcome',['name'=>'test'] )
~~~
配置文件在 `lang/zh-cn/app.php`中。
搜索替换\n , ，空格。 转成数组 
~~~
string_to_array($str)
~~~

AES解密
~~~
$token = $_GET['token']; 
$token = aes_decode($token);
pr($token);
~~~

根据用户id查用户信息
~~~
get_user($user_id)
~~~
查用户信息，where条件
~~~
get_user_where($where = [])
~~~

取用户扩展字段值
~~~
get_user_meta($user_id)
~~~
更新用户的meta信息
 ，参数$meta如  ['nickname'=>'']
~~~
set_user_meta($user_id,$meta)
~~~


获取HTTP_REFERER
~~~
referer($show_full = false)
~~~

当前域名 ，结尾带/
~~~
host()
~~~

取用户扩展字段值
~~~
get_user_meta_where($where = [],$return_row = false)
~~~


所有会员
~~~
get_user_all($where = [])
~~~


取单个用户组信息
~~~
user_group_get($group_id)
~~~

跳转
~~~
jump($url)
~~~

CDN地址,需要在`config.ini.php`配置` $config['cdn_url'] = [];`
~~~
static_url()
~~~

判断是命令行下
~~~
 is_cli()
~~~

判断是否为json 
~~~
is_json($str)
~~~

数组转对象
~~~
array_to_object($arr)
~~~

对象转数组
~~~
object_to_array($obj)
~~~

取目录名
~~~
get_dir($name)
~~~

取后缀，不包含`.`
~~~
get_ext($name)
~~~

取文件名
~~~
get_name($name)
~~~
创建目录 
~~~
create_dir($arr = [])
~~~
取IP
~~~
get_ip()
~~~
计算两点地理坐标之间的距离,返回公里数
~~~
get_distance($longitude1, $latitude1, $longitude2, $latitude2)
~~~
设置、获取cookie
~~~
cookie($name, $value = null, $expire = 0, $path = '/')
~~~
删除COOKIE 
~~~
remove_cookie($name)
~~~
返回  min="1900-01-01" max="2099-12-31"
~~~
date_limit()
~~~
路径列表，支持文件夹下的子所有文件夹
~~~
get_deep_dir($path)
~~~
返回成功信息，JSON格式 
~~~
json_success($arr = [])
~~~
返回错误信息，JSON格式 
~~~
json_error($arr = [])
~~~
yaml转数组，数组转yaml格式
~~~
yaml($str)
~~~
解析xlsx文件
~~~
$lists = lib\Xls::load($file, [
            '产品编号' => 'product_num',
            '产品规格' => 'name',
            '注册证号' => 'cert_num',
            '单位'     => 'unit',
        ],$column_use_date = []); 
~~~
`$column_use_date ` 指定哪几列是时间格式
xls生成
~~~
生成多个worksheet事例：
$where = [
    'status'    => 1,
    'date[>=]' => $start,
    'date[<=]' => $end,
];
$title  = [
      'invoice_number'   => '发票号码',
      'customer_name'    => '供应商',
      'product_num'      => '规格型号',
      'num'              => '数量',
      'price'            => '单价',
      'total_price'      => '金额',
]; 
//专票
$where['invoice_type'] = 1;
$values = invoice_set_xls_data($where,$title);
//普票
$where['invoice_type'] = 2;
$new_data = invoice_set_xls_data($where,$title);
//第一个worksheet
Xls::$label = '专票';
Xls::$sheet_width = [
    'A' => "15",
    'B' => "36",
    'C' => "30",
    'D' => "10",
    'E' => "10",
    'F' => "10",
];
//更多worksheet，如果只是一个，可不用下面代码，直接跳到Xls::create处
Xls::$works = [
    [
        'title' => $title,
        'label' => '普票',
        'data'  => $new_data,
        'width' => Xls::$sheet_width,
    ]
];
Xls::create($title, $values, $name, true);
~~~
验证数据
~~~
$data    = g();   
$vali    = validate([
    'company_title'   => '客户名',
    'email'   => '邮件地址',
    'active_plugins'  => '系统',
    'exp_time' => '过期时间',
],$data,[
    'required' => [
        ['company_title'],
        ['email'],
        ['active_plugins'],
        ['exp_time'],
    ],
    'email'=>[
        ['email']
    ]
]);
if($vali){
    json($vali);
} 
~~~
规则 
~~~
required - Field is required
requiredWith - Field is required if any other fields are present
requiredWithout - Field is required if any other fields are NOT present
equals - Field must match another field (email/password confirmation)
different - Field must be different than another field
accepted - Checkbox or Radio must be accepted (yes, on, 1, true)
numeric - Must be numeric
integer - Must be integer number
boolean - Must be boolean
array - Must be array
length - String must be certain length
lengthBetween - String must be between given lengths
lengthMin - String must be greater than given length
lengthMax - String must be less than given length
min - Minimum
max - Maximum
listContains - Performs in_array check on given array values (the other way round than in)
in - Performs in_array check on given array values
notIn - Negation of in rule (not in array of values)
ip - Valid IP address
ipv4 - Valid IP v4 address
ipv6 - Valid IP v6 address
email - Valid email address
emailDNS - Valid email address with active DNS record
url - Valid URL
urlActive - Valid URL with active DNS record
alpha - Alphabetic characters only
alphaNum - Alphabetic and numeric characters only
ascii - ASCII characters only
slug - URL slug characters (a-z, 0-9, -, _)
regex - Field matches given regex pattern
date - Field is a valid date
dateFormat - Field is a valid date in the given format
dateBefore - Field is a valid date and is before the given date
dateAfter - Field is a valid date and is after the given date
contains - Field is a string and contains the given string
subset - Field is an array or a scalar and all elements are contained in the given array
containsUnique - Field is an array and contains unique values
creditCard - Field is a valid credit card number
instanceOf - Field contains an instance of the given class
optional - Value does not need to be included in data array. If it is however, it must pass validation.
arrayHasKeys - Field is an array and contains all specified keys.
~~~

数组转tree 
~~~
/**
 * 数组转tree 
 * 
 * 输入$list 
 * [
 *   {id:1,pid:0,其他字段},
 *   {id:2,pid:1,其他字段},
 *   {id:3,pid:1,其他字段},
 * ]
 * 输出 
 * [
 *   [
 *      id:1,
 *      pid:0,
 *      其他字段,
 *      children:[
 *           {id:2,pid:1,其他字段},
 *           {id:3,pid:1,其他字段},
 *      ]
 *   ]
 * ]
 * 
 */
array_to_tree($list, $pk = 'id', $pid = 'pid', $child = 'children', $root = 0, $my_id = '')
~~~

### 插件

在`plugins`目录创建`backrestoreapi`

info.php 代码

~~~
<?php 

return [
	'title'  => '为开发提供数据接口恢复',  
	'version'=> '1.0.1', 
	'level' => 100,
	'data'  => [
		'desc' => '',
		'price'=> '',
		'free' => 0, 
		'dependents'=> [
			 
		],
	]
];
~~~

include.php 代码

~~~
add_action("admin.menu",function(&$menu)
{
 	$menu['backrestoreapi'] = [
		'icon'=>'',
        'label'=>'网站开发接口恢复', 
 	];
 	$menu['backrestoreapi']['children'][] = [
 		'icon'=>'far fa-circle',
        'label'=>'应用',
        'url'=>'plugins/backrestoreapi/app.php', 
        'acl'=>[ 
        ]
 	];
});
~~~

### 自动退款、当使用pay插件时有用

refund_all.php

~~~
<?php 

include __DIR__.'/../app.php'; 

if(!is_cli()){
    echo "403";exit;
}
//防止重复执行 
cli_prevent_duplication($argv, $cmd = 'php refund_all.php');
echo "对所有支付进行退款\n";

$all = db_get("pay","*",['status[!]'=>-100]);
$pay = new plugins\pay\lib\WxPayBase(get_config('wx_appid')); 
if(!$all){
	echo "无需退款";
	exit;
}
foreach($all as $v){
	$out_trade_no = $v['out_trade_no'];
	$transaction_id = $v['transaction_id'];
	$total_fee = $v['total_fee'];
	$status = $v['status'];
	if($transaction_id && $total_fee){
		$r = $pay->refund($out_trade_no,$total_fee,$total_fee,$refund_desc = '退款');
		if($r['code'] == 0){
			echo "退款成功 ".$total_fee."\n";
			db_update("pay",['status'=>-100],['id'=>$v['id']]);
		}
	}
}
~~~

### 数据库

#### HOOk

~~~
//写入数据前
do_action("db_insert.$table.before",$data); 
//写入数据后
do_action("db_insert.$table.after",$id); 
//更新数据前
do_action("db_update.$table.before",$data); 
//更新数据后
do_action("db_update.$table.after",$data); 
//查寻数据
do_action("db_get_one.$table",$one);  
//删除数据前
do_action("db_insert.$table.del",$where); 
~~~


#### 数据库事务
~~~

db()->action(function() { 
    db_insert("aa",[  
        'amount'=>1,  
        'real_amount'=>1,  
        'order_num'=>1,  
        'invoicemoney_get_id'=>1,
        'type' => 'bank',
        'hexiao_status'=>1, 
        'created_at'=>now(), 
    ]);  
    db_insert("aa",[ 
        'real_amount'=>1,  
        'order_num'=>1,  
        'invoicemoney_get_id'=>1,
        'type' => 'bank',
        'hexiao_status'=>1, 
        'created_at'=>now(), 
    ]);  
    $err = db_get_error(); 
    if($err){ 
        return false;    
    } 
});
exit;
~~~
如果使用ThinkOrm，事务代码如下所示
~~~
use think\facade\Db;
Db::startTrans();
try { 
        Db::table("aa")->insert([  
            'amount'=>1,  
            'real_amount'=>1,  
            'order_num'=>1,  
            'invoicemoney_get_id'=>1,
            'type' => 'bank',
            'hexiao_status'=>1, 
            'created_at'=>now(), 
        ]);  
        Db::table("aa")->insert([ 
            'real_amount'=>1,  
            'order_num'=>1,  
            'invoicemoney_get_id'=>1,
            'type' => 'bank',
            'hexiao_status'=>1, 
            'created_at'=>now(), 
        ]);  
        // 提交事务
        Db::commit();  
} catch (\Exception $e) {
    pr($e->getMessage());
    Db::rollback();
} 
exit; 
~~~


**使用Medoo数据库操作类**

具体用法请查看。
https://medoo.in/

数据库配置在`config.ini.php`中。
在使用时可直接
~~~
db()->select(table,field,where)
~~~
如 https://medoo.in/api/select
不同的是这里`db()` 是已经初始化数据库对象了。
**读取表中的字段**
~~~ 
get_table_fields($table)
~~~
**返回数据库允许的数据，传入其他字段自动忽略**
此方法常用于过滤掉表不需要的字段。
~~~
$data = db_allow($table,$data);
~~~
显示数据库表结构，支持markdown格式
~~~
database_tables($name = null,$show_markdown = false)
~~~

**分页查寻**
~~~
JOIN
 $where = [ 
    //"do_order.id"=>1,
    'ORDER'=>[
        'do_order.id'=>'DESC'
    ]
]; 

int date
$where['printer_refund_apply.created_at[<>]']  = [
 $dates[0] / 1000, $dates[1] / 1000
];
datetime
$where['printer_refund_apply.created_at[<>]']  = [
 date('Y-m-d H:i:s',$dates[0] / 1000), date('Y-m-d H:i:s',$dates[1] / 1000)
];

$data = db_pager("do_order",
    ["[><]do_mini_user" => ["uid" => "id"]],
    [
        "do_order.id",
        "do_order.uid",
        "user" => [
            "do_mini_user.nickName",
            "do_mini_user.avatarUrl",
            "do_mini_user.openid",
        ]
    ],
    $where);

~~~

**根据表名、字段 、条件 查寻多条记录** 
~~~
db_get($table, $join = "*", $columns=null, $where=null)
~~~

**写入记录**
~~~
db_insert($table, $data = [])
~~~

**更新记录**
~~~
db_update($table, $data = [], $where = [])
~~~
 

**根据表名、字段 、条件 查寻一条记录**
~~~
db_get_one($table, $join  = "*", $columns=null, $where=null)
~~~
**执行SQL**
~~~ 
db_query($sql,$raw=null)  //$raw有参数时为数组传值 
~~~
如
~~~
db_query("select * from user where user=:user",[":user"=>'admin'])
~~~ 

**取最小值**
~~~
db_get_min($table, $join  = "*", $column=null, $where=null)
~~~
**取最大值**
~~~
db_get_max($table, $join =  "*", $column = null, $where = null)
~~~
**总数**
~~~
db_get_count($table, $join =  "*", $column = null, $where = null)
~~~
是否有记录
~~~
db_get_has($table, $join = null, $where = null)
~~~
随机取多条记录
~~~
db_get_rand($table, $join= "*", $column=null, $where=null)
~~~
取总和
~~~
db_get_sum($table, $join="*", $column=null, $where=null)
~~~
取平均值
~~~
db_get_avg($table, $join="*", $column=null, $where=null)
~~~
删除
~~~
db_del($table, $where)
db_delete($table, $where)
~~~
显示所有表名
~~~
show_tables($table)
~~~
$where 条件：AND OR 查寻 https://medoo.in/api/where
~~~
 "AND" => [
    "OR" => [
        "user_name" => "foo",
        "email" => "foo@bar.com"
    ],
    "password" => "12345"
]
 ~~~
SQL：
~~~
WHERE (user_name = 'foo' OR email = 'foo@bar.com') AND password = '12345'
~~~
$where 条件：更复杂的AND OR 查寻
~~~
[
    "AND #Actually, this comment feature can be used on every AND and OR relativity condition" => [
        "OR #the first condition" => [
            "user_name" => "foo",
            "email" => "foo@bar.com"
        ],
        "OR #the second condition" => [
            "user_name" => "bar",
            "email" => "bar@foo.com"
        ]
    ]
]
~~~
SQL：
~~~
WHERE (
    ("user_name" = 'foo' OR "email" = 'foo@bar.com')
    AND
    ("user_name" = 'bar' OR "email" = 'bar@foo.com')
)
~~~
**连接新的数据库**
~~~
多数据库时有用
$config['db_host'] = '127.0.0.1';
//数据库名
$config['db_name'] = 'dbname';
//数据库登录用户名
$config['db_user'] = 'root';
//数据库登录密码
$config['db_pwd']  = '111111'; 
//数据库端口号
$config['db_port'] = 3306;
 
$new_db = new_db($config = []);
~~~
用返回的`$new_db `进行操作。
如
~~~
$new_db->insert(table,data);
~~~
用法与Medoo一样。


changelog.md
### v1.0.3

优化db，使用 composer require catfan/medoo



inc/db.php
<?php

/*
	Copyright (c) 2021-2050 FatPlug, All rights reserved.
	This file is part of the FatPlug Framework (http://fatplug.cn).
	This is not free software.
	you can redistribute it and/or modify it under the
	terms of the License after purchased commercial license. 
	mail: sunkangchina@163.com
	web: http://fatplug.cn
*/

if (!defined('VERSION')) {
    die();
}
/**
复杂的查寻，(...  AND ...) OR (...  AND ...)
"OR #1" => [
    "AND #2" => $where,
    "AND #3" => $or_where
    ]
];  

 */
global $db;
global $_db_par;
//连接数据库  
try {
    $pdo = new PDO($dsn, $user, $pwd);
    $db = new Medoo\Medoo([
        'pdo'     => $pdo,
        'type'    => 'mysql',
        'option'  => [
            PDO::ATTR_CASE => PDO::CASE_NATURAL
        ],
        'command' => [
            'SET SQL_MODE=ANSI_QUOTES'
        ],
        'error' => PDO::ERRMODE_WARNING
    ]);
} catch (Exception $e) {
    $err = $e->getMessage();
    write_log_error("连接数据库失败");
    add_action("db_connect_error", $err);
}

/**
多数据库时有用
$config['db_host'] = '127.0.0.1';
//数据库名
$config['db_name'] = 'saas_admin';
//数据库登录用户名
$config['db_user'] = 'root';
//数据库登录密码
$config['db_pwd']  = '111111'; 
//数据库端口号
$config['db_port'] = 3306;
 */
function new_db($config = [])
{
    $db = new \Medoo\Medoo([
        'type' => 'mysql',
        'host' => $config['db_host'],
        'database' => $config['db_name'],
        'username' => $config['db_user'],
        'password' => $config['db_pwd'],
        // [optional]
        'charset'   => 'utf8mb4',
        'collation' => 'utf8mb4_general_ci',
        'port'      => $config['db_port'],
        'prefix'    => '',
        'error'     => PDO::ERRMODE_SILENT,
        // Read more from http://www.php.net/manual/en/pdo.setattribute.php.
        'option'    => [
            PDO::ATTR_CASE => PDO::CASE_NATURAL
        ],
        'command' => [
            'SET SQL_MODE=ANSI_QUOTES'
        ]
    ]);
    return $db;
}
/**
 * 数据库实例
 *
 * @return object
 */
function db()
{
    global $db;
    return $db;
}

/***
 * 分页查寻 
 JOIN
 $where = [ 
    //"do_order.id"=>1,
    'ORDER'=>[
        'do_order.id'=>'DESC'
    ]
]; 

int date
$where['printer_refund_apply.created_at[<>]']  = [
 $dates[0] / 1000, $dates[1] / 1000
];
datetime
$where['printer_refund_apply.created_at[<>]']  = [
 date('Y-m-d H:i:s',$dates[0] / 1000), date('Y-m-d H:i:s',$dates[1] / 1000)
];

$data = db_pager("do_order",
    ["[><]do_mini_user" => ["uid" => "id"]],
    [
        "do_order.id",
        "do_order.uid",
        "user" => [
            "do_mini_user.nickName",
            "do_mini_user.avatarUrl",
            "do_mini_user.openid",
        ]
    ],
    $where);

 * @param string $table 表名
 * @param string $column 字段 
 * @param array $where  条件 [LIMIT=>1]  
 * @return array
 */


function db_pager_count($nums = null)
{
    static $_page_count;
    if ($nums && $nums >= 0) {
        $_page_count = $nums;
    } else {
        return $_page_count;
    }
}

function db_pager($table, $join, $columns = null, $where = null)
{
    global $_db_par;
    $flag = true;
    if (!$where) {
        $where   = $columns;
        $columns = $join ?: "*";
        $join    = '';
        $count   = db_pager_count() ?: db_get_count($table, $where);
    } else if ($join && $where) {
        $flag    = false;
        $count   = db_pager_count() ?: db_get_count($table, $join, "$table.id", $where);
    }
    $current_page  = $_REQUEST['page'];
    if (!$current_page) {
        $current_page  = g('page');
    }
    if (!$current_page) {
        $current_page = 1;
    }
    $p1 = $_REQUEST['per_page'] ?: g('per_page');
    $p2 = $_REQUEST['page_size'] ?: g('page_size');
    if ($p1 >= 1) {
        $pre_page  = $p1;
    } else {
        $pre_page  = $p2;
    }
    if (!$pre_page) {
        $pre_page = 20;
    }
    $count = (int)$count;
    $last_page     = ceil($count / $pre_page);
    $has_next_page = $last_page > $current_page ? true : false;
    $start         = ($current_page - 1) * $pre_page;
    $where['LIMIT'] = [$start, $pre_page];
    if ($flag) {
        $data  =  db_get($table, $columns, $where);
    } else {
        $data  =  db_get($table, $join, $columns, $where);
    }
    $_db_par['size'] = $pre_page;
    $_db_par['count'] = $count;
    return [
        'current_page' => $current_page,
        'last_page'    => $last_page,
        'pre_page'     => $pre_page,
        'total'        => $count,
        'has_next_page' => $has_next_page,
        'data'         => $data,
        'data_count'   => count($data),
    ];
}
/**
 * 显示分页
 * 调用db_pager后，再调用。
 * db_pager_html([
 *      'url'=>'',
 * ]);
 */
function db_pager_html($arr = [])
{
    global $_db_par;
    if ($arr['count']) {
        $count  = $arr['count'];
    } else {
        $count = $_db_par['count'];
    }
    $page_url = $arr['url'];
    if ($arr['size']) {
        $size  = $arr['size'];
    } else {
        $size = $_db_par['size'] ?: 20;
    }
    $paginate = new \lib\Paginate($count, $size);
    if ($page_url) {
        $paginate->url = $page_url;
    }
    $limit  = $paginate->limit;
    $offset = $paginate->offset;
    return $paginate->show();
}

global $db_error;
function db_add_error($str)
{
    global $db_error;
    $db_error[] = $str;
}

function db_get_error()
{
    global $db_error;
    if ($db_error)
        return $db_error;
}

/***
 * 根据表名、字段 、条件 查寻多条记录
 * where https://medoo.in/api/where
 * select($table, $columns)
 * select($table, $columns, $where)
 * select($table, $join, $columns, $where)
 * @param string $table 表名
 * @param string $column 字段 
 * @param array $where  条件 [LIMIT=>1]  
 * @return array
 */
function db_get($table, $join = null, $columns = null, $where = null)
{
    if (is_array($join)) {
        foreach ($join as $k => $v) {
            if (is_string($v) && strpos($v, '(') !== false) {
                $join[$k] = db_raw($v);
            }
        }
    }
    if (is_string($columns) && strpos($columns, 'WHERE') !== FALSE) {
        $columns = db_raw($columns);
    }
    try {
        $all =  db()->select($table, $join, $columns, $where);
        //查寻数据
        do_action("db_get.$table", $all);
        return $all;
    } catch (Exception $e) {
        db_add_error($e->getMessage());
    }
}
/** 
$lists = db_select('do_order', [ 
            'count' => 'COUNT(`id`)',
            'total' => 'SUM(`total_fee`)',
            'date'  => "FROM_UNIXTIME(`inserttime`, '%Y-%m-%d')"
        ], 
        'WHERE `status` = 1 GROUP BY `date` LIMIT 30'
);  
 */
function db_select($table, $join = "*", $columns = null, $where = null)
{
    return db_get($table, $join, $columns, $where);
}
/**
 * 写入记录
 *
 * @param string $table 表名 
 * @param array  $data  数据 
 * @return array
 */
function db_insert($table, $data = [])
{
    foreach ($data as $k => $v) {
        if (substr($k, 0, 1) == "_") {
            unset($data[$k]);
        }
    }
    try {
        //写入数据前
        do_action("db_insert.$table.before", $data);
        $db    = db()->insert($table, $data);
        $id = db()->id();
        //写入数据后
        do_action("db_insert.$table.after", $id);
        return $id;
    } catch (Exception $e) {
        db_add_error($e->getMessage());
    }
}

/**
 * 写入记录
 *
 * @param string $table 表名 
 * @param array  $data  数据 
 * @return array
 */
function db_update($table, $data = [], $where = [])
{
    global $db_where;
    $db_where = $where;

    foreach ($data as $k => $v) {
        if (substr($k, 0, 1) == "_") {
            unset($data[$k]);
        }
    }
    try {
        //更新数据前
        do_action("db_update.$table.before", $data);
        $db    = db()->update($table, $data, $where);
        $error = db()->error;
        if ($error) {
            write_log(['db_error' => $error, 'sql' => db()->last()]);
            throw new Exception($error);
        }
        $count =  $db->rowCount();
        //更新数据后
        do_action("db_update.$table.after", $data);
        return $count;
    } catch (Exception $e) {
        db_add_error($e->getMessage());
    }
}

/**
 * 数据库事务
 *
 */
function action($call)
{
    $result = "";
    $db     = db();
    $db->action(function ($db) use (&$result, $call) {
        $call();
    });
}
/**
 * 根据表名、字段 、条件 查寻一条记录
 *
 * @param string $table 表名
 * @param string $column 字段 
 * @param array  $where 条件 
 * @return array
 */
function db_get_one($table, $join  = "*", $columns = null, $where = null)
{
    if (!$where) {
        $columns['LIMIT'] = 1;
    } else {
        $where['LIMIT']   = 1;
    }
    $db = db_get($table, $join, $columns, $where);
    if ($db) {
        $one =  $db[0];
        //查寻数据
        do_action("db_get_one.$table", $one);
        return $one;
    }
    return;
}
/**
 * 执行SQL
 *
 * @param string $sql
 * @return void
 */
function sql_run($sql)
{
    db()->query($sql);
}

/**
 * SQL查寻
 */
function db_query($sql, $raw = null)
{
    if ($raw === null) {
        return db()->query($sql);
    }
    $q = db()->query($sql, $raw);
    if ($q) {
        $all =  $q->fetchAll(\PDO::FETCH_ASSOC) ?: [];
        //查寻数据
        do_action("db_query", $all);
        return $all;
    } else {
        return [];
    }
}
/**
 * Quotes the string for the query.
 * https://medoo.in/api/quote
 */
function db_quote($data)
{
    return db()->quote($data);
}

/**
 * 取最小值 
 * https://medoo.in/api/min
 * min($table, $column, $where)
 * min($table, $join, $column, $where)
 * @param string $table  表名
 * @param string $column 字段 
 * @param array $where   条件
 * @return void
 */
function db_get_min($table, $join  = "*", $column = null, $where = null)
{
    return db()->min($table, $join, $column, $where);
}

/**
 * 取最大值  
 * max($table, $column, $where)
 * max($table, $join, $column, $where)
 * @param string $table  表名
 * @param string $column 字段 
 * @param array $where   条件
 * @return void
 */
function db_get_max($table, $join =  "*", $column = null, $where = null)
{
    return db()->max($table, $join, $column, $where);
}

/**
 * 总数  
 * count($table, $where)
 * count($table, $join, $column, $where)
 * @param string $table  表名 
 * @param array $where   条件
 * @return void
 */
function db_get_count($table, $join =  "*", $column = null, $where = null)
{
    return db()->count($table, $join, $column, $where);
}

/**
 * 是否有记录
 * has($table, $where)
 * has($table, $join, $where)
 * @param string $table  表名 
 * @param array $where   条件
 * @return void
 */
function db_get_has($table, $join = null, $where = null)
{
    return db()->has($table, $join, $where);
}

/**
 * 随机取多条记录  
 * rand($table, $column, $where)
 * rand($table, $join, $column, $where)
 * @param string $table  表名
 * @param string $column 字段 
 * @param array $where   条件
 * @return void
 */
function db_get_rand($table, $join = "*", $column = null, $where = null)
{
    return db()->rand($table, $join, $column, $where);
}

/**
 * 取总和
 * sum($table, $column, $where)
 * sum($table, $join, $column, $where)
 * @param string $table  表名
 * @param string $column 字段 
 * @param array $where   条件
 * @return void
 */
function db_get_sum($table, $join = "*", $column = null, $where = null)
{
    return db()->sum($table, $join, $column, $where);
}

/**
 * 取平均值 
 * avg($table, $column, $where)
 * avg($table, $join, $column, $where)
 * @param string $table  表名
 * @param string $column 字段 
 * @param array $where   条件
 * @return void
 */
function db_get_avg($table, $join = "*", $column = null, $where = null)
{
    return db()->avg($table, $join, $column, $where);
}

/**
 * RAW
 * https://medoo.in/api/raw
 * raw('NOW()')
 * raw('RAND()')
 * raw('AVG(<age>)') 
 * @param string $raw
 * @return  
 */
function db_raw($raw)
{
    return \Medoo\Medoo::raw($raw);
}

//删除
function db_del($table, $where)
{
    //删除数据前
    do_action("db_insert.$table.del", $where);
    return db()->delete($table, $where);
}

function db_delete($table, $where)
{
    return db_del($table, $where);
}

/**
 * 显示所有表名
 *
 * @param string $table 表名
 * @version 1.0.0
 * @author sun <sunkangchina@163.com>
 * @return void
 */
function show_tables($table)
{
    $sql = "SHOW TABLES LIKE '%$table%';";
    $all = db()->query($sql);
    foreach ($all as $v) {
        foreach ($v as $v1) {
            $list[] = $v1;
        }
    }
    return $list;
}


/**
 * 读取表中的字段
 * 在正式环境有缓存, fields:
 * @param string $table
 * @return void
 */
function allow_fields($table, $has_key  = true)
{
    $sql   = "SHOW FULL FIELDS FROM " . $table . "";
    $lists = db()->query($sql);
    $arr   = [];
    foreach ($lists as $vo) {
        if ($has_key) {
            $arr[$vo['Field']] = $vo['Field'];
        } else {
            $arr[] = $vo['Field'];
        }
    }
    return $arr;
}

/**
 * 取表中字段
 */
function get_table_fields($table, $flag = true)
{
    return allow_fields($table, $flag);
}
/**
 *返回数据库允许的数据，传入其他字段自动忽略
 */
function db_allow($table, $data)
{
    return get_data_db_allow($table, $data);
}
function get_data_db_allow($table, $data)
{
    $fields = get_table_fields($table);
    foreach ($data as $k => $v) {
        if (!$fields[$k]) {
            unset($data[$k]);
        }
    }
    return $data;
}

/**
 * 显示数据库表结构，支持markdown格式
 * @param string $name 数据库名
 * @version 1.0.0
 * @author sun <sunkangchina@163.com>
 */
function database_tables($name = null, $show_markdown = false)
{
    global $config;
    if (!$name) {
        $name = $config['dbname'];
    }
    $sql  = "SHOW TABLE STATUS FROM {$name}";
    $all  = db_query($sql, []);
    foreach ($all as $k => $v) {
        $sql   = "SHOW FULL FIELDS FROM " . $v['Name'] . "";
        $lists = db()->query($sql, []);
        $all[$k]['FIELDS'] = $lists;
    }
    if (!$show_markdown) {
        return $all;
    }
    $str = "";
    foreach ($all as $v) {
        $str .= "###### " . $v['Name'] . " " . $v['Comment'] . "\n";
        $str .= "| 字段  |  类型 | 备注|\n";
        $str .= "| ------------ | ------------ |------------ |\n";
        foreach ($v['FIELDS'] as $vo) {
            $str .= "|  " . $vo['Field'] . " |  " . $vo['Type'] . " |" . $vo['Comment'] . "|\n";
        }
        $str .= "\n\n";
    }
    return $str;
}
